tools/ktx/command_info.cpp

```diff
@@ -0,0 +1,242 @@
+        <dd>Specifies the output format.
+            @b text - Human readable text based format.
+            @b json - Formatted JSON.
+            @b mini-json - Minified JSON (Every optional formatting is skipped).
+            The default format is @b text.
+        </dd>
+    </dl>
+    @snippet{doc} ktx/command.h command options
+
+@section ktxtools_info_exitstatus EXIT STATUS
+    @b ktx @b info exits
+        0 on success,
+        1 on command line errors and
+        2 if the input file parsing failed.
+
+@section ktxtools_info_history HISTORY
+
+@par Version 4.0
+ - Initial version
+
+@section ktxtools_info_author AUTHOR
+    - Mátyás Császár [Vader], RasterGrid www.rastergrid.com
+    - Daniel Rákos, RasterGrid www.rastergrid.com
+*/
+
+class CommandInfo : public Command {
+    struct Options {
+        enum class OutputFormat {
+            text,
+            json,
+            json_mini,
+        };
+
+        OutputFormat format = OutputFormat::text;
+        _tstring inputFilepath;
+    };
+
+    Options options;
+
+public:
+    // TODO KTX Tools P5: Support --version
+    // TODO KTX Tools P5: Support --help with proper usage
+    void initializeOptions();
+    virtual bool processOption(argparser& parser, int opt) override;
+    void processPositional(const std::vector<_tstring>& infiles, const _tstring& outfile);
+    virtual int main(int argc, _TCHAR* argv[]) override;
+
+    using Command::Command;
+    virtual ~CommandInfo() {};
+
+private:
+    int printInfoText(const _tstring& infile);
+    int printInfoJSON(const _tstring& infile, bool minified);
+};
+
+// -------------------------------------------------------------------------------------------------
+
+void CommandInfo::initializeOptions() {
+    option_list.emplace(option_list.begin(), "format", argparser::option::required_argument, nullptr, 'f');
+    short_opts += "f:";
+}
+
+bool CommandInfo::processOption(argparser& parser, int opt) {
+    switch (opt) {
+    case 'f':
+        if (parser.optarg == "text") {
+            options.format = Options::OutputFormat::text;
+        } else if (parser.optarg == "json") {
+            options.format = Options::OutputFormat::json;
+        } else if (parser.optarg == "mini-json") {
+            options.format = Options::OutputFormat::json_mini;
+        } else {
+            // TODO KTX Tools P5: Print usage, Failure: unsupported format
+            std::cerr << "Print usage, Failure: unsupported format" << std::endl;
+            return false;
+        }
+        break;
+    default:
+        return false;
+    }
+
+    return true;
+}
+
+void CommandInfo::processPositional(const std::vector<_tstring>& infiles, const _tstring& outfile) {
+    if (infiles.empty()) {
+        // TODO KTX Tools P5: Print usage, Failure: infile is missing
+        std::cerr << "Print usage, Failure: infile is missing" << std::endl;
+        // return EXIT_FAILURE;
+        return;
+    }
+    if (infiles.size() > 1) {
```

Comment:
Yes. In fact for the info and validate commands it would be weird to take multiple inputs, as for JSON output it would either mean outputting multiple JSONs or we would need to adjust the agreed JSON format to be arrayed.

Also, I think in particular for the validate command it's useful to not support multiple inputs as a caller could use the result code of the program to determine the success of validation of individual files if they are required to be issued separately.

In general, the project specification requires a single input file in all cases, except for create where multiple inputs may be baked into a single output.

Reply:
I can see how it might be a problem for the JSON output. Would outputting a line with the file name before the JSON output break the typical JSON use cases?

Re the validate command, the current command is silent when the file is valid, simply exiting with status 0. It prints the name of a failing file at the top of the diagnosis. In either case, therefore, processing multiple files is not an issue. If a user wants success or failure for individual files they can call validate with a single file. In other words they can write a shell script. Something they would have to do when processing multiple files if the tool only accepts a single input. So I can't see that supporting multiple inputs adds any burden to that use case.

Reply:
> Would outputting a line with the file name before the JSON output break the typical JSON use cases?

Yes. Output must be a valid JSON document. One way to address this is to adjust the root JSON object to be an array, another is to handle one input at a time.

Reply:
In general, I'd prefer to not support multiple inputs on the tool level, except when they are implied by a command.

External batching via shell scripts is more flexible anyway and also simplifies the CLI tool development and future maintenance.

Reply:
> In general, I'd prefer to not support multiple inputs on the tool level, except when they are implied by a command.
> External batching via shell scripts is more flexible anyway and also simplifies the CLI tool development and future maintenance.

I agree with @lexaknyazev.

Unless we make the JSON root an array, or an object with the filename as key, or some other additional level, most JSON parsers will not be able to handle multiple separate JSONs in the output (with or without filenames in between). That means that any consumer would have to pre-parse the output in order to be able to extract individual JSONs.

@MarkCallow, is there any particular reason why you'd like to deviate from the project specification here and allow multiple input files?

Reply:
Only the long history of Unix commands. However it is clear this is an issue for the JSON output so I agree we should stick with the original spec.